### ISSUE TITLE
[ Question ] Write throughput, performance?

### ISSUE THREAD

**Issue by Samusername** [zalando/patroni#1395]
I have tried to test performance of writes.
Write throughput is quite low, in my setup, it seems, in comparison to one other alternative. ( ? ) 

I am not certain do I have something incorrectly somewhere.

Write thoughput is very good with one other postgres HA solution in k8s.
With patroni I can reach one half / one third (or less) of write throughput rate, in comparison to the other one. Measurement: rows per second.
I installed patroni M-R-R cluster with help of Zalando postgres-operator.

Table / one row has:
bigint, small text, timestamp, and jsonb data 
(jsonb: quite small entries, like 12 - 25 fields with values in one entry).

Write throughput, rate: 13 000 obj / s.

Master - sync replica - async replica
( I have tried also with Master - async - async, but it did not seem to affect so much. )

I use same data generator in both cases: in patroni case and the other HA solution case.
In both alternatives:
- Same data generator
- Same HW, in same k8s
- PostgreSQL is in use. 
- Replication: M - sync replica - async replica. sync/async states are visible in: pg_stat_replication.
streaming replication.
- shared_buffers: couple of GBs
- svc:
  patroni, k8s svc: LoadBalancer, 
  other HA solution, k8s svc: NodePort 

I will try to repeat the tests / check settings, etc., in case I did something incorrectly.

Any ideas for settings which I should try to modify and check/try to improve the write throughput? In patroni (or Zalando postgres-operator).

Has someone else tested the write throughputs? Results?

( The other alternative, which I have tried, is Stolon. )

(I do not know is this patroni or Zalando postgres-operator with patroni topic.)

**Comment by CyberDem0n** [comment]
If you are using the same hardware performance should be comparable. There is probably some difference in the configuration of postgres or your ingress setup (LoadBalancer vs NodePort).

You will have to figure where is the bottleneck on your own.

**Comment by Samusername** [comment]
I created the M-R-R cluster again (with help of Zalando postgres-operator).
I created NodePort instead of LoadBalancer now.
But the write throughput is still the same. It did not improve.

I access the DB through the NodePort / svc.
( I am not sure do I find other possible ways to improve the performance / write throughput. )

**Comment by CyberDem0n** [comment]
Have accurately read my message?
Especially this part "There is probably some difference in the configuration of postgres".
Have you tried to compare all postgres configuration parameters?

**Comment by Samusername** [comment]
It is a great hint! Now I understood. Thanks! 
I am checking the differences in postgres configurations. ( "show all;" )

I tried to turn off archive_mode, but that did not have effects on the write performance.

Differences, which I found, are listed next. (In case someone has hints what could be changed.)

patroni vs stolon

 **archive_command                     | [-/bin/true-] {+(disabled)+}                              
 archive_mode                        | [-on-] {+off+}                                     
 archive_timeout                     | [-30min-] {+0+}** 

 **data_checksums                      | [-on-] {+off+}**  


List of entires, enable_, Not in stolon:

 **enable_parallel_append                 | on        
 enable_parallel_hash                   | on    
 enable_partition_pruning               | on     
 enable_partitionwise_aggregate         | off 
 enable_partitionwise_join              | off**  

Individual lines,

 **[-extwlist.custom_path                   | /scripts                   
 extwlist.extensions                    | btree_gin,btree_gist,citext,hstore,intarray,ltree,pgcrypto,pgq,pg_trgm,postgres_fdw,uuid-ossp,hypopg,timescaledb,pg_partman-]** 

List of entries, jit_, not in stolon:

 **jit_above_cost                         | 100000                     
 jit_debugging_support                  | off                        
 jit_dump_bitcode                       | off                        
 jit_expressions                        | on                         
 jit_inline_above_cost                  | 500000                     
 jit                                    | off                        
 jit_optimize_above_cost                | 500000                     
 jit_profiling_support                  | off                        
 jit_provider                           | llvmjit                    
 jit_tuple_deforming                    | on**       

log_:

 **log_autovacuum_min_duration         | [-0-] {+-1+}                                      
 log_checkpoints                     | [-on-] {+off+}                                     
 log_connections                     | [-on-] {+off+}                                     
 log_destination                     | [-csvlog-] {+stderr+}                                  
 log_directory                       | [-../pg_log-] {+log+}                                     
 log_disconnections                  | [-on-] {+off+}    
...
 log_lock_waits                      | [-on-] {+off+}    
 log_min_duration_statement          | [-500ms-] {+-1+}                                      
 log_statement                       | [-all-] {+none+}                                    
 log_temp_files                      | [-0-] {+-1+}**  

 **ssl                                    [-| on**  

In patroni:
 ssl                                    | on        
 ssl_passphrase_command                 |                                                         
 ssl_passphrase_command_supports_reload | off    

In stolon:
 ssl                                 | off   
 ssl_prefer_server_ciphers           | on   

tcp_:
 **tcp_keepalives_count                | [-0-] {+9+}                                       
 tcp_keepalives_idle                 | [-0-] {+7200+}                                    
 tcp_keepalives_interval             | [-0-] {+75+}** 

 **[-timescaledb.disable_load               | off                                           
 timescaledb.max_background_workers     | 8-]**    

 **wal_log_hints                       | [-on-] {+off+}**  

I will try to check how to turn off e.g. ssl (if possible).

I could try to change settings one by one, but I would be very grateful of any hints what should It try! (if anything)


All differencies,
with grepping the differences with egrep "\[|\]", from output from dwdiff, patroni vs stolon:

```
 archive_command                     | [-/bin/true-] {+(disabled)+}                              
 archive_mode                        | [-on-] {+off+}                                     
 archive_timeout                     | [-30min-] {+0+}                                       
 autovacuum_analyze_scale_factor     | [-0.02-] {+0.1+}                                     
 autovacuum_max_workers              | [-5-] {+3+}                                       
 autovacuum_vacuum_scale_factor      | [-0.05-] {+0.2+}                                     
 [-bg_mon.listen_address                  | ::                                                                                                                          
 bg_mon.port                            | 8080-]                                                                                                                        
 checkpoint_completion_target        | [-0.9-] {+0.5+}                                     
 cluster_name                        | [-acid-test-cluster-]                                                                                                           
 config_file                         | [-/home/postgres/pgdata/pgroot/data/postgresql.conf-] {+/stolon-data/postgres/postgresql.conf+}   
 [-cron.database_name                     | postgres                                                                                                                    
 cron.max_running_jobs                  | 32-]                                                                                                                          
 data_checksums                      | [-on-] {+off+}                                     
 data_directory                      | [-/home/postgres/pgdata/pgroot/data                                                                                           
 data_sync_retry                        | off-] {+/stolon-data/postgres+}                   
 effective_cache_size                | [-7000MB-] {+4GB+}                                     
 [-enable_parallel_append                 | on                                                                                                                          
 enable_partitionwise_join              | off-]                                                                                                                         
 [-extwlist.custom_path                   | /scripts                                                                                                                    
 extwlist.extensions                    | btree_gin,btree_gist,citext,hstore,intarray,ltree,pgcrypto,pgq,pg_trgm,postgres_fdw,uuid-ossp,hypopg,timescaledb,pg_partman-] 
 hba_file                            | [-/home/postgres/pgdata/pgroot/data/pg_hba.conf-] {+/stolon-data/postgres/pg_hba.conf+}       
 ident_file                          | [-/home/postgres/pgdata/pgroot/data/pg_ident.conf-] {+/stolon-data/postgres/pg_ident.conf+}     
 [-jit_above_cost                         | 100000                                                                                                                      
 jit_tuple_deforming                    | on-]                                                                                                                          
 lc_collate                          | [-en_US.UTF-8-] {+en_US.utf8+}                              
 lc_ctype                            | [-en_US.UTF-8-] {+en_US.utf8+}                              
 lc_messages                         | [-en_US.UTF-8-] {+en_US.utf8+}                              
 lc_monetary                         | [-en_US.UTF-8-] {+en_US.utf8+}                              
 lc_numeric                          | [-en_US.UTF-8-] {+en_US.utf8+}                              
 lc_time                             | [-en_US.UTF-8-] {+en_US.utf8+}                              
 listen_addresses                    | [-*-] {+127.0.0.1,192.168.133.4+}                 
 log_autovacuum_min_duration         | [-0-] {+-1+}                                      
 log_checkpoints                     | [-on-] {+off+}                                     
 log_connections                     | [-on-] {+off+}                                     
 log_destination                     | [-csvlog-] {+stderr+}                                  
 log_directory                       | [-../pg_log-] {+log+}                                     
 log_disconnections                  | [-on-] {+off+}                                     
 log_file_mode                       | [-0644-] {+0600+}                                    
 log_filename                        | [-postgresql-%u.log-] {+postgresql-%Y-%m-%d_%H%M%S.log+}          
 logging_collector                   | [-on-] {+off+}                                     
 log_line_prefix                     | [-%t [%p]: [%l-1] %c %x %d %u %a %h-] {+%m [%p]+}                                 
 log_lock_waits                      | [-on-] {+off+}                                     
 log_min_duration_statement          | [-500ms-] {+-1+}                                      
 log_statement                       | [-all-] {+none+}                                    
 log_temp_files                      | [-0-] {+-1+}                                      
 log_timezone                        | [-Etc/UTC-] {+UTC+}                                     
 log_truncate_on_rotation            | [-on-] {+off+}                                     
 [-max_parallel_maintenance_workers       | 2-]                                                                                                                           
 max_replication_slots               | [-10-] {+20+}                                      
 max_wal_senders                     | [-10-] {+47+}                                      
 [-parallel_leader_participation          | on-]                                                                                                                          
 [-pg_stat_kcache.linux_hz                | 1000000                                                                                                                     
 pg_stat_statements.track_utility       | off-]                                                                                                                         
 server_version                      | [-11.6 (Ubuntu 11.6-1.pgdg18.04+1)-] {+10.5 (Debian 10.5-2.pgdg90+1)+}           
 server_version_num                  | [-110006-] {+100005+}                                  
 [-set_user.block_alter_system            | on                                                                                                                          
 set_user.superuser_whitelist           | *-]                                                                                                                           
 shared_buffers                      | [-6000MB-] {+3GB+}                                     
 shared_preload_libraries            | [-bg_mon,pg_stat_statements,pgextwlist,pg_auth_mon,set_user,timescaledb,pg_cron,pg_stat_kcache-]                                
 ssl_cert_file                       | [-/home/postgres/server.crt-]                                                                                                   
 ssl_key_file                        | [-/home/postgres/server.key-]                                                                                                   
 ssl                                    [-| on                                                                                                                          
 ssl_passphrase_command_supports_reload-] | off                                     
 synchronous_standby_names           | [-"acid-test-cluster-2"-] {+stolon_1c7c6f2e+}                         
 tcp_keepalives_count                | [-0-] {+9+}                                       
 tcp_keepalives_idle                 | [-0-] {+7200+}                                    
 tcp_keepalives_interval             | [-0-] {+75+}                                      
 [-timescaledb.disable_load               | off                                                                                                                         
 timescaledb.max_background_workers     | 8-]                                                                                                                           
 TimeZone                            | [-Etc/UTC-] {+UTC+}                                     
 track_activity_query_size           | [-1kB-] {+1024+}                                    
 track_functions                     | [-all-] {+none+}                                    
 unix_socket_directories             | [-/var/run/postgresql-] {+/tmp+}                                    
 [-vacuum_cleanup_index_scale_factor      | 0.1-]                                                                                                                         
 wal_keep_segments                   | [-300-] {+8+}                                       
 wal_log_hints                       | [-on-] {+off+}                                     

```

**Comment by Samusername** [comment]
I changed, e.g.:

... checkpoint_completion_target: 0.9 --> 0.1
And:
use_pg_rewind: true --> false

But those do not seem to affect much.


It was not possible to disable data_checksums in manifest, through 
```
postgresql:
  parameters:
    data_checksums: false
```

When pod / postgres tries to start up:
```
Error:
2020-02-19 11:22:33.126 UTC [15518] LOG:  parameter "data_checksums" cannot be changed
2020-02-19 11:22:33 UTC [15518]: [2-1] 5e4d1a79.3c9e 0     FATAL:  configuration file "/home/postgres/pgdata/pgroot/data/postgresql.conf" contains errors
```

**Comment by Samusername** [comment]
I was not able to set initdb: ... data-checksums: false, with help of manifest. 
Error occurred in cluster creation: 
2020-02-19 12:11:49,248 INFO: trying to bootstrap a new cluster
/usr/lib/postgresql/11/bin/initdb: option '--data-checksums' doesn't allow an argument

**Comment by CyberDem0n** [comment]
I would rather think that is is related to the log_statement=all than anything else.

**Comment by Samusername** [comment]
That was it! Now the write performance is similar (same) with patroni in comparison to stolon. 

I used value log_statement: "none" in this test.

I actually searched from internet information about log_statement already "at first", but there was some page shown where it was more like recommended to use the "all" setting. It was mentioned: 'all' "should be your default". So, I did not try to change it right away. So, it actually seems to affect dramatically to the write throughput (like common sense may possibly indicate).
